### PR TITLE
handles integrity error in enterprise permissions

### PR DESCRIPTION
**Pull request by Robert-Costello**
## Summary
[USH-778](https://dimagi-dev.atlassian.net/browse/USH-778)

## Feature Flag
COVID: Enterprise Permissions

## Product Description
If a user attempts to add a project space that has already been added, they are presented with a message indicating that they have already added that particular project space. 

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan
none

### Safety story
Change should only affect the scope of this particular feature flag (Enterprise Permissions)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/users/views/__init__.py [hunk 1]
```diff
@@ -704,10 +705,15 @@ def create_domain_permission_mirror(request, domain):
...
         mirror_domain = Domain.get_by_name(mirror_domain_name)
         if mirror_domain is not None:
-            mirror = DomainPermissionsMirror(source=domain, mirror=mirror_domain_name)
-            mirror.save()
-            message = _('You have successfully added the project space "{mirror_domain_name}".')
-            messages.success(request, message.format(mirror_domain_name=mirror_domain_name))
+            try:
+                mirror = DomainPermissionsMirror(source=domain, mirror=mirror_domain_name)
+                mirror.save()
+            except IntegrityError:
```
Comment (esoergel) [inline corehq/apps/users/views/__init__.py:711]: nit: this check might make more sense in `clean_mirror_domain` as part of form validation.

Comment (esoergel) [reply corehq/apps/users/views/__init__.py:711]: actually same with the `if mirror_domain is not None` check above, though i know you didn't add that.  By moving them there, the errors would show up attached to the input field, which also seems nicer.

**Review by esoergel — APPROVED** [review]
I think the validation should be moved to the form (see comments in-line), but that's not blocking.

File: corehq/apps/users/views/__init__.py [hunk 2]
```diff
@@ -704,10 +705,15 @@ def create_domain_permission_mirror(request, domain):
...
+                message = _(mirror_domain_name + ' has already been added.')
```
Comment (orangejenny) [inline corehq/apps/users/views/__init__.py:712]: This needs to use `format` like the error message below. If you pass a dynamic string to `_`, the calculated string isn't going to have a translation, so it won't get translated. [More about translations](https://commcare-hq.readthedocs.io/translations.html).

Comment (Robert-Costello) [reply corehq/apps/users/views/__init__.py:712]: Understood, just updated it to use format.

Comment (orangejenny) [reply corehq/apps/users/views/__init__.py:712]: I don't see a change - possibly forgot to push changes?

Comment (Robert-Costello) [reply corehq/apps/users/views/__init__.py]: Sorry, yeah I was looking at Ethan's notes and forgot to push. Should be up now.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/users/views/__init__.py (hunk 1 above)
Comment (Robert-Costello) [reply corehq/apps/users/views/__init__.py:711]: I see what you mean. I'll look more at swapping those checks over tomorrow.

Comment (Robert-Costello) [reply corehq/apps/users/views/__init__.py]: I moved those error checks over to clean_mirror_domain. It seems to work with no issues when I test it locally, but if you want to take another peak I've pushed up the changes.

File: corehq/apps/users/views/__init__.py [hunk 3]
```diff
@@ -701,25 +701,13 @@ def delete_domain_permission_mirror(request, domain, mirror):
...
-            else:
-                message = _('You have successfully added the project space "{mirror_domain_name}".')
-                messages.success(request, message.format(mirror_domain_name=mirror_domain_name))
-        else:
-            message = _('Please enter a valid project space.')
-            messages.error(request, message.format())
-    else:
+    if not form.is_valid():
         for field, message in form.errors.items():
             messages.error(request, message)
```
Comment (esoergel) [inline corehq/apps/users/views/__init__.py:705]: I didn't notice this earlier - looks like this isn't a standard CRUD operation where it re-renders the form with errors, it instead logs them as messages, so my previous comment about getting in-line errors was wrong.  Ah well.

File: corehq/apps/users/forms.py [hunk 4]
```diff
@@ -1273,17 +1274,26 @@ def clean_domains(self):
...
+        mirror_domain = Domain.get_by_name(mirror_domain_name)
+        if mirror_domain is not None:
+            try:
+                mirror = DomainPermissionsMirror(source=self.domain, mirror=mirror_domain)
+                mirror.save()
+            except IntegrityError:
+                message = '"' + mirror_domain_name + '"' + ' has already been added.'
+                raise forms.ValidationError(_(message))
+        else:
+            raise forms.ValidationError(_('Please enter valid project space.'))
```
Comment (esoergel) [inline corehq/apps/users/forms.py:1298]: sorry for being vague earlier - I do think the save action belongs elsewhere, probably in the main view or in a `save` method on the form (either is fine).  Expectation would be that `form.is_valid` only checks validity, and doesn't yet commit any changes.

You could do something like this here, then create the mirror in the view:
```suggestion
        mirror_domain = Domain.get_by_name(mirror_domain_name)
        if not mirror_domain:
            raise forms.ValidationError(_('Please enter valid project space.'))
        if DomainPermissionsMirror.objects.filter(mirror=mirror_domain).exists():
            message = _('"{mirror_domain_name}" has already been added.')
            raise forms.ValidationError(message.format(mirror_domain_name=mirror_domain_name))
```
(it looks like you accidentally reverted the translation change Jenny mentioned - added that back too)

If you're concerned about needing to duplicate the db constraint checks, a `ModelForm` would add that validation to `is_valid` automatically, though that'd take a bit more rejiggering of this view and form, so I probably wouldn't bother.

**Review by esoergel — CHANGES_REQUESTED** [review]

File: corehq/apps/users/forms.py (hunk 4 above)
Comment (Robert-Costello) [reply corehq/apps/users/forms.py]: No worries, thanks for bearing with me while I get familiar with the code. I made these changes.

**Review by esoergel — APPROVED** [review]
thanks for the patience!

**Pull request merged**